scripts/apm-server-load-test.js

```diff
@@ -215,52 +229,119 @@ async function postPayload(url, payload) {
-    },
-    body: data.join('')
-  })
-  return response.text()
+
+  try {
+    let response = await fetch(url, {
+      method: 'POST',
+      cache: 'no-cache',
+      headers: {
+        'Content-Type': 'application/x-ndjson',
+        'X-Forwarded-For': faker.internet.ip(),
+        'User-Agent': faker.internet.userAgent()
+      },
+      body: data.join('')
+    })
+    return response.text()
+  } catch (error) {
+    return error
+  }
 }
 
-async function generatePayloads(transactionCount) {
+function asyncTimeout(ms) {
+  return new Promise(resolve => setTimeout(resolve, ms))
+}
+async function generatePayloads(transactionCount, trPerSession = 10) {
   const spanPerTransaction = 50
   let promises = []
+  let sessionId
+  let baseUrl
+  let payloads = []
   for (let i = 0; i < transactionCount; i++) {
-    let payload = generateTransaction(spanPerTransaction)
+    const sequence = i % trPerSession
+    if (sequence == 0) {
+      sessionId = generateRandomId(16)
+      baseUrl = faker.internet.url()
+    }
+
+    let payload = generateTransaction(
+      spanPerTransaction,
+      sessionId,
+      sequence + 1,
+      baseUrl
+    )
     payload.unshift(merge({}, defaultMeta))
+    payloads.push(payload)
     let p = postPayload(`${serverUrl}/intake/v2/rum/events`, payload)
     promises.push(p)
   }
-  await Promise.all(promises)
-  let response = await fetch(`${serverUrl}/debug/vars`)
-  const apmServerResults = await response.json()
+  let transactionResponses = await Promise.all(promises)
+  const responses = {}
+  transactionResponses.forEach(tr => {
+    if (!responses[tr]) {
+      responses[tr] = 1
+    } else {
+      responses[tr]++
+    }
+  })
 
   const result = {
     '@timestamp': Date.now(),
     transactionCount,
     spanPerTransaction,
-    apmServer: apmServerResults
+    apmServer: null,
+    transformStats: null,
+    responses
+  }
+
+  try {
+    let response = await fetch(`${serverUrl}/debug/vars`)
+    const apmServerResults = await response.json()
+    result.apmServer = apmServerResults
+  } catch (error) {
+    result.apmServer = error
   }
+
+  if (debug) {
+    console.log(responses)
+    result.payloads = payloads
+    try {
+      let transformStats = await fetch(
+        `${esUrl}/_transform/transform_rum_sessions/_stats`,
+        {
+          method: 'GET',
+          headers: {
+            Authorization: esAuth
+          }
+        }
+      )
+      result.transformStats = await transformStats.json()
+    } catch (error) {
+      result.transformStats = error
+    }
+  }
+
   return result
 }
+const iterations = 10
 
 ;(async function () {
   const outputFile = process.argv[2]
-
-  const result = await generatePayloads(10)
+  const results = []
+  for (let iteration = 0; iteration < iterations; iteration++) {
+    if (debug) {
+      console.log(`Iteration: ${iteration}`)
+    }
+    const result = await generatePayloads(100)
+    results.push(result)
+    await asyncTimeout(1000)
```

Comment:
is this for simulating load? may be give it a random number between `1-1000ms` as well for delay?

Reply:
This is actually to avoid hitting the rate limit, otherwise requests will be rejected